Actions/DownloadProjectDependencies/DownloadProjectDependencies.Action.ps1

```diff
@@ -84,45 +78,59 @@ function DownloadDependenciesFromCurrentBuild($baseFolder, $project, $projectsDe
+    Write-Host "::endgroup::"
 
-Write-Host "::group::Downloading project dependencies from probing paths"
-$downloadedDependencies += DownloadDependenciesFromProbingPaths -baseFolder $baseFolder -project $project -destinationPath $destinationPath
-Write-Host "::endgroup::"
+    Write-Host "::group::Downloading project dependencies from probing paths"
+    $settings = $env:Settings | ConvertFrom-Json | ConvertTo-HashTable -recurse
+    $downloadedDependencies += DownloadDependenciesFromProbingPaths -settings ([ref]$settings) -baseFolder $baseFolder -project $project -destinationPath $destinationPath
+    Write-Host "::endgroup::"
 
-Write-Host "Downloaded dependencies: $($downloadedDependencies -join ', ')"
+    Write-Host "Downloaded dependencies: $($downloadedDependencies -join ', ')"
 
-$downloadedApps = @()
-$downloadedTestApps = @()
+    $downloadedApps = @()
+    $downloadedTestApps = @()
 
-# Split the downloaded dependencies into apps and test apps
-$downloadedDependencies | ForEach-Object {
-    # naming convention: app, (testapp)
-    if ($_.startswith('(')) {
-        $DownloadedTestApps += $_    
-    }
-    else {
-        $DownloadedApps += $_    
+    # Split the downloaded dependencies into apps and test apps
+    $downloadedDependencies | ForEach-Object {
+        # naming convention: app, (testapp)
+        if ($_.startswith('(')) {
+            $DownloadedTestApps += $_
+        }
+        else {
+            $DownloadedApps += $_
+        }
     }
-}
 
-Write-Host "Downloaded dependencies apps: $($DownloadedApps -join ', ')"
-Write-Host "Downloaded dependencies test apps: $($DownloadedTestApps -join ', ')"
+    Write-Host "Downloaded dependencies apps: $($DownloadedApps -join ', ')"
+    Write-Host "Downloaded dependencies test apps: $($DownloadedTestApps -join ', ')"
+
+    $DownloadedAppsJson = ConvertTo-Json $DownloadedApps -Depth 99 -Compress
+    $DownloadedTestAppsJson = ConvertTo-Json $DownloadedTestApps -Depth 99 -Compress
 
-$DownloadedAppsJson = ConvertTo-Json $DownloadedApps -Depth 99 -Compress
-$DownloadedTestAppsJson = ConvertTo-Json $DownloadedTestApps -Depth 99 -Compress
+    Add-Content -Encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "DownloadedApps=$DownloadedAppsJson"
+    Add-Content -Encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "DownloadedTestApps=$DownloadedTestAppsJson"
 
-Add-Content -Encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "DownloadedApps=$DownloadedAppsJson"
-Add-Content -Encoding UTF8 -Path $env:GITHUB_OUTPUT -Value "DownloadedTestApps=$DownloadedTestAppsJson"
+    'appFolders', 'testFolders', 'bcptTestFolders' | ForEach-Object {
+        $propName = $_
+        Write-Host "Setting $($propName) to $($settings."$propName" -join ', ')"
+        $foldersJson = ConvertTo-Json $settings."$propName" -Depth 99 -Compress
+        Add-Content -Encoding UTF8 -Path $env:GITHUB_ENV -Value "$propName=$foldersJson"
```

Comment:
Because appDependencyProbingPaths supports 'include' - meaning that you can include apps from other projects in a project.
So, artifact is read in ReadSettings and placed as ENV var - potentially overridden in DetermineArtifactUrl
appFolders, testFolders and bcptTestFolders are read in ReadSettings and potentially overridden in DownloadProjectDependencies.

If we don't do this, then we cannot resolve appDependencyProbingPaths in the action (or only part of it at least)

Reply:
Let me know if you have other ideas - I don't think that partially solving appDependencyProbingPaths is the right way - then we should move all back to RunPipeline, which also doesn't seem right

Reply:
Then ReadSettings becomes a huge function.

Here's how it works now:
ReadSettings should read the settings and settings.artifact should contain the value of artifact given in settings for this branch, this workflow etc. With the get parameter, we request to have artifact set as environment variable as well.
Then DetermineArtifactUrl will decipher the artifact setting and set env:artifact to the actual URL to use including the insider sas token. (in the old days this action would re-write env:settings)
RunPipeline then gets env:artifact transferred as parameter.

If we were to resolve artifact in ReadSettings we would spend time on querying artifacts store etc. in order to get the right artifactURL - ReadSettings would also need to get secrets and getting secrets needs to get settings.

The same pattern is implemented for appFolders, which are affected by appDependencyProbingPaths, which sometimes needs secrets.

Enumerating appDependencyProbingPaths while reading settings seems wrong.

So now, env:appFolders are the folders that needs to be built and settings.appFolders are the folders you have in settings  - always.